CRM/Price/BAO/PriceSet.php

```diff
@@ -1602,15 +1602,17 @@ public static function copyPriceSet($baoName, $id, $newId) {
    *
    * @return array
    */
-  public static function setLineItem($field, $lineItem, $optionValueId, &$totalTax) {
+  public static function setLineItem($field, &$lineItem, $optionValueId, &$totalTax) {
     // Here we round - i.e. after multiplying by quantity
+    $taxRate = $field['options'][$optionValueId]['tax_rate'];
+    $unitPrice = (($lineItem[$optionValueId]['unit_price'] * 100) / (100 + $taxRate));
+
     if ($field['html_type'] == 'Text') {
-      $taxAmount = round($field['options'][$optionValueId]['tax_amount'] * $lineItem[$optionValueId]['qty'], 2);
+      $taxAmount = round(($unitPrice * $lineItem[$optionValueId]['qty'] * ($taxRate / 100)), 2);
     }
     else {
-      $taxAmount = round($field['options'][$optionValueId]['tax_amount'], 2);
+      $taxAmount = round(($unitPrice * ($taxRate / 100)), 2);
```

Comment:
I'm not sure we should be rounding at the BAO layer, it should be happening at the form layer / before saving calculated values to DB.  Also, we should be using functions in CRM_Utils_Money where possible which allows us to change the number of decimal places in future should a currency require it.

Reply:
Yeah we need to be really careful here as it is centralised code and it will affect lots of flows